stan/math/gpu/opencl_context.hpp

```diff
@@ -303,30 +303,36 @@ class opencl_context {
       msg << "Platform Vendor: " << platform.getInfo<CL_PLATFORM_VENDOR>()
           << "\n";
 
-      std::vector<cl::Device> all_devices;
-      platform.getDevices(CL_DEVICE_TYPE_GPU, &all_devices);
+      try {
+        std::vector<cl::Device> all_devices;
+        platform.getDevices(CL_DEVICE_TYPE_GPU, &all_devices);
 
-      for (auto device_iter : all_devices) {
-        cl::Device device(device_iter);
+        for (auto device_iter : all_devices) {
+          cl::Device device(device_iter);
 
-        msg << "\tDevice " << device_id++ << ": "
-            << "\n";
-        msg << "\t\tDevice Name: " << device.getInfo<CL_DEVICE_NAME>() << "\n";
-        msg << "\t\tDevice Type: " << device.getInfo<CL_DEVICE_TYPE>() << "\n";
-        msg << "\t\tDevice Vendor: " << device.getInfo<CL_DEVICE_VENDOR>()
-            << "\n";
-        msg << "\t\tDevice Max Compute Units: "
-            << device.getInfo<CL_DEVICE_MAX_COMPUTE_UNITS>() << "\n";
-        msg << "\t\tDevice Global Memory: "
-            << device.getInfo<CL_DEVICE_GLOBAL_MEM_SIZE>() << "\n";
-        msg << "\t\tDevice Max Clock Frequency: "
-            << device.getInfo<CL_DEVICE_MAX_CLOCK_FREQUENCY>() << "\n";
-        msg << "\t\tDevice Max Allocateable Memory: "
-            << device.getInfo<CL_DEVICE_MAX_MEM_ALLOC_SIZE>() << "\n";
-        msg << "\t\tDevice Local Memory: "
-            << device.getInfo<CL_DEVICE_LOCAL_MEM_SIZE>() << "\n";
-        msg << "\t\tDevice Available: " << device.getInfo<CL_DEVICE_AVAILABLE>()
-            << "\n";
+          msg << "\tDevice " << device_id++ << ": "
+              << "\n";
+          msg << "\t\tDevice Name: " <<
+            device.getInfo<CL_DEVICE_NAME>() << "\n";
+          msg << "\t\tDevice Type: " <<
+            device.getInfo<CL_DEVICE_TYPE>() << "\n";
+          msg << "\t\tDevice Vendor: " <<
+            device.getInfo<CL_DEVICE_VENDOR>() << "\n";
+          msg << "\t\tDevice Max Compute Units: "
+              << device.getInfo<CL_DEVICE_MAX_COMPUTE_UNITS>() << "\n";
+          msg << "\t\tDevice Global Memory: "
+              << device.getInfo<CL_DEVICE_GLOBAL_MEM_SIZE>() << "\n";
+          msg << "\t\tDevice Max Clock Frequency: "
+              << device.getInfo<CL_DEVICE_MAX_CLOCK_FREQUENCY>() << "\n";
+          msg << "\t\tDevice Max Allocateable Memory: "
+              << device.getInfo<CL_DEVICE_MAX_MEM_ALLOC_SIZE>() << "\n";
+          msg << "\t\tDevice Local Memory: "
+              << device.getInfo<CL_DEVICE_LOCAL_MEM_SIZE>() << "\n";
+          msg << "\t\tDevice Available: " <<
+              device.getInfo<CL_DEVICE_AVAILABLE>() << "\n";
+        }
+      } catch (const cl::Error& e) {
+        msg << "\tno GPU devices in the specified platform" << "\n";
```

Comment:
Nice catch! Instead of having a message here could we make a call to [`check_opencl_error()`](https://github.com/stan-dev/math/pull/720/files#diff-c10b73fa94faecc3f0dec24a8c80b261R166)? I think it would throw a `-1001: "CL_PLATFORM_NOT_FOUND_KHR` or if not we can figure out what error code it returns and setup an error code for it.

Reply:
Should we throw at all since having additional platforms without GPUs is not a problem and could happen?

We are just outputing the information here. 

This is only a problem if the user selecta the platform that has no GPU devices. But that is handled elsewhere and does throw an exception, as it should :)

Reply:
Should we throw at all since having additional platforms without GPUs is not a problem and could happen?

We are just outputing the information here. 

This is only a problem if the user selecta the platform that has no GPU devices. But that is handled elsewhere and does throw an exception, as it should :)

Reply:
I think if we can throw a more informative error message then "C++ exception with description "clGetPlatformIDs" than it's a good idea to catch and handle it